// @vtex/react-csv-parse/lib/index.js
'use strict';

exports.__esModule = true;

var _splitAt = require('ramda/src/splitAt');

var _splitAt2 = _interopRequireDefault(_splitAt);

var _zipObj = require('ramda/src/zipObj');

var _zipObj2 = _interopRequireDefault(_zipObj);

var _lift = require('ramda/src/lift');

var _lift2 = _interopRequireDefault(_lift);

var _apply = require('ramda/src/apply');

var _apply2 = _interopRequireDefault(_apply);

var _compose = require('ramda/src/compose');

var _compose2 = _interopRequireDefault(_compose);

var _dropLast = require('ramda/src/dropLast');

var _dropLast2 = _interopRequireDefault(_dropLast);

var _last = require('ramda/src/last');

var _last2 = _interopRequireDefault(_last);

var _isEmpty = require('ramda/src/isEmpty');

var _isEmpty2 = _interopRequireDefault(_isEmpty);

var _splitEvery = require('ramda/src/splitEvery');

var _splitEvery2 = _interopRequireDefault(_splitEvery);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

// NOTES
// Separator is the character splitting the data. Comma ex.: 1,2,3
// Delimiter is a text delimiter. Double quotes ex.: "Account", "Id"

var CsvParse = function (_React$Component) {
  _inherits(CsvParse, _React$Component);

  function CsvParse() {
    var _temp, _this, _ret;

    _classCallCheck(this, CsvParse);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, _React$Component.call.apply(_React$Component, [this].concat(args))), _this), _this.handleOnChange = function (event) {
      var file = event.target.files[0];
      _this.handleSeparators(file);
    }, _temp), _possibleConstructorReturn(_this, _ret);
  }

  CsvParse.prototype.handleSeparators = function handleSeparators(file) {
    var _this2 = this;

    var reader = new FileReader();
    reader.readAsText(file);
    reader.onload = function () {
      var result = reader.result;
      var splitResult = result.split('');

      // check if first character is not alphanumeric
      if (!splitResult[0].match(/^[a-z0-9]+$/i, 'g')) {
        // get index of second delimiter
        var secondDelimiterIndex = result.indexOf(splitResult[0], 1);

        // separator is character after second delimiter
        var separator = result[secondDelimiterIndex + 1];

        // clean text delimiters
        var delimiter = splitResult[0];
        var dynamicRegEx = new RegExp(delimiter, 'g');
        result = result.replace(dynamicRegEx, '');

        // start parsing
        _this2.parseData(result, separator);
      } else {
        // first character is alphanum, guessing the separator
        _this2.props.separators.some(function (val, i) {
          var dynamicRegEx = new RegExp(_this2.props.separators[i], 'g');
          var separatorsFound = (result.match(dynamicRegEx) || []).length;
          if (separatorsFound >= _this2.props.keys.length) {
            _this2.parseData(result, _this2.props.separators[i]);
            return true;
          } else if (i === _this2.props.separators.length - 1) {
            _this2.props.onDataUploaded(null);
          }
        });
      }
    };
  };

  CsvParse.prototype.parseData = function parseData(result, separator) {
    // replace line breaks and tabs by separator
    result = result.replace(/[\r\n\t]/g, separator);

    // split string by separator
    result = result.split(separator);

    // create arrays at each headers' length string
    result = (0, _splitEvery2.default)(this.props.keys.length, result);

    // drop last item if empty
    if ((0, _isEmpty2.default)((0, _last2.default)(result)[0])) {
      result = (0, _dropLast2.default)(1, result);
    }

    // remove display headers
    result.shift();

    // add api headers
    result.unshift(this.props.keys);

    // convert arrays to objects
    result = (0, _compose2.default)((0, _apply2.default)((0, _lift2.default)(_zipObj2.default)), (0, _splitAt2.default)(1))(result);

    // save it all in state
    this.props.onDataUploaded(result);
  };

  CsvParse.prototype.render = function render() {
    return this.props.render(this.handleOnChange);
  };

  return CsvParse;
}(_react2.default.Component);

CsvParse.defaultProps = {
  separators: [',', ';']
};

CsvParse.propTypes = process.env.NODE_ENV !== "production" ? {
  keys: _propTypes2.default.array.isRequired,
  onDataUploaded: _propTypes2.default.func.isRequired,
  separators: _propTypes2.default.array
} : {};

exports.default = CsvParse;
module.exports = exports['default'];
